# Assignment 3

## Question 1: 

### Apply user-based collaborative filtering to the course topics dataset. All recomendations will be 1. Explain why this is.



In [90]:
import pandas as pd 

data = pd.read_csv("Coursetopics.csv")
data

Intro  DataMining  Survey  Cat Data  Regression  Forecast  DOE  SW
0        1           1       0         0           0         0    0   0
1        0           0       1         0           0         0    0   0
2        0           1       0         1           1         0    0   1
3        1           0       0         0           0         0    0   0
4        1           1       0         0           0         0    0   0
..     ...         ...     ...       ...         ...       ...  ...  ..
360      0           0       0         1           0         0    0   0
361      0           1       0         1           0         0    0   1
362      0           0       0         0           0         0    0   1
363      0           0       0         1           0         0    0   0
364      0           0       0         0           1         0    0   0

[365 rows x 8 columns]

In [91]:
data.shape

(365, 8)

In [92]:
data.isna().sum()

Intro         0
DataMining    0
Survey        0
Cat Data      0
Regression    0
Forecast      0
DOE           0
SW            0
dtype: int64

In [93]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori,association_rules

frequent_itemsets = apriori(data,min_support=0.005,use_colnames=True) 

rules = association_rules(df=frequent_itemsets,metric="confidence",min_threshold=0.8)

rules[["antecedents","consequents","confidence","lift"]]

antecedents          consequents  \
0                            (Regression, DOE)              (Intro)   
1                (Regression, DOE, DataMining)              (Intro)   
2                         (Survey, DOE, Intro)           (Cat Data)   
3               (Survey, Regression, Forecast)              (Intro)   
4                     (Survey, Regression, SW)              (Intro)   
5                       (Survey, SW, Forecast)              (Intro)   
6                  (Regression, DOE, Cat Data)              (Intro)   
7                        (Regression, SW, DOE)              (Intro)   
8                    (Survey, DOE, DataMining)           (Cat Data)   
9               (Survey, Regression, Forecast)         (DataMining)   
10               (Regression, DOE, DataMining)           (Cat Data)   
11                 (Forecast, DOE, DataMining)           (Cat Data)   
12                     (Survey, DOE, Forecast)           (Cat Data)   
13                      (Survey, SW, Forecast)           (Cat Data)   
14                    (Survey, Regression, SW)                (DOE)   
15     (Forecast, Intro, Cat Data, DataMining)             (Survey)   
16    (Survey, Cat Data, Forecast, DataMining)              (Intro)   
17            (Survey, DOE, Intro, DataMining)           (Cat Data)   
18     (Survey, Regression, Intro, DataMining)           (Forecast)   
19  (Survey, Regression, Forecast, DataMining)              (Intro)   
20       (Survey, Regression, Forecast, Intro)         (DataMining)   
21              (Survey, Regression, Forecast)  (Intro, DataMining)   
22        (Regression, DOE, Intro, DataMining)           (Cat Data)   
23     (Regression, DOE, Cat Data, DataMining)              (Intro)   
24               (Regression, DOE, DataMining)    (Cat Data, Intro)   
25           (SW, Intro, Cat Data, DataMining)         (Regression)   
26       (Survey, Regression, Cat Data, Intro)                (DOE)   
27         (Survey, Regression, DOE, Cat Data)              (Intro)   
28              (Survey, DOE, Forecast, Intro)           (Cat Data)   
29            (Forecast, DOE, Cat Data, Intro)             (Survey)   
30               (Survey, SW, Forecast, Intro)           (Cat Data)   
31             (Forecast, SW, Cat Data, Intro)             (Survey)   
32            (Survey, SW, Cat Data, Forecast)              (Intro)   
33                      (Survey, SW, Forecast)    (Cat Data, Intro)   
34             (Survey, Regression, SW, Intro)                (DOE)   
35               (Survey, Regression, SW, DOE)              (Intro)   
36                    (Survey, Regression, SW)         (DOE, Intro)   
37             (Regression, SW, Cat Data, DOE)              (Intro)   

    confidence       lift  
0     0.818182   2.073864  
1     1.000000   2.534722  
2     0.800000   3.842105  
3     1.000000   2.534722  
4     1.000000   2.534722  
5     1.000000   2.534722  
6     1.000000   2.534722  
7     1.000000   2.534722  
8     1.000000   4.802632  
9     1.000000   5.615385  
10    1.000000   4.802632  
11    1.000000   4.802632  
12    1.000000   4.802632  
13    1.000000   4.802632  
14    1.000000   5.793651  
15    1.000000   5.367647  
16    1.000000   2.534722  
17    1.000000   4.802632  
18    1.000000   7.156863  
19    1.000000   2.534722  
20    1.000000   5.615385  
21    1.000000  18.250000  
22    1.000000   4.802632  
23    1.000000   2.534722  
24    1.000000  14.038462  
25    1.000000   4.802632  
26    1.000000   5.793651  
27    1.000000   2.534722  
28    1.000000   4.802632  
29    1.000000   5.367647  
30    1.000000   4.802632  
31    1.000000   5.367647  
32    1.000000   2.534722  
33    1.000000  14.038462  
34    1.000000   5.793651  
35    1.000000   2.534722  
36    1.000000  21.470588  
37    1.000000   2.534722

One reason why the confidence might be 1 for every rule is that every time we see the antecedent, we see the consequent, and vice versa. In short, every course that is related to another course appears when the other course also appears, making the confidence essentially equal to 100%. 

## Question 2:

### The file Cosmetics.csv contains the data on the cosmetics purchases. Using Python, apply association rules to these data (use min_support=0.1 and min_threshold=0.8).

In [94]:
import pandas as pd
cos_data = pd.read_csv("Cosmetics.csv")
cos_data.head()

Trans.   Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  \
0        1    0      1            1        1          1                0   
1        2    0      0            1        0          1                0   
2        3    0      1            0        0          1                1   
3        4    0      0            1        1          1                0   
4        5    0      1            0        0          1                0   

   Bronzer  Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  \
0        1          1        1           0           0          0         0   
1        1          1        0           0           1          1         0   
2        1          1        1           1           1          1         1   
3        1          0        0           0           1          0         0   
4        1          1        1           1           0          1         1   

   Eyeliner  
0         1  
1         0  
2         0  
3         1  
4         0

In [95]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori,association_rules

cos_data = cos_data.drop("Trans. ",axis=1)
frequent_itemsets = apriori(cos_data,min_support=0.1,use_colnames=True) 

rules = association_rules(df=frequent_itemsets,metric="confidence",min_threshold=0.8)

rules[["antecedents","consequents","confidence","lift"]]

antecedents    consequents  confidence      lift
0                           (Brushes)  (Nail Polish)    1.000000  3.571429
1                        (Eye shadow)      (Mascara)    0.842520  2.359999
2                           (Mascara)   (Eye shadow)    0.899160  2.359999
3                  (Lip liner, Blush)    (Concealer)    0.870968  1.970515
4                 (Eye shadow, Blush)      (Mascara)    0.928571  2.601040
5                    (Mascara, Blush)   (Eye shadow)    0.918478  2.410704
6           (Eye shadow, Nail Polish)      (Mascara)    0.908397  2.544529
7              (Mascara, Nail Polish)   (Eye shadow)    0.888060  2.330865
8                (Lip liner, Bronzer)    (Concealer)    0.804687  1.820560
9                 (Eyeliner, Bronzer)    (Concealer)    0.815068  1.844046
10              (Eyeliner, Lip liner)    (Concealer)    0.923077  2.088409
11            (Concealer, Eye shadow)      (Mascara)    0.890547  2.494530
12               (Concealer, Mascara)   (Eye shadow)    0.877451  2.303021
13              (Eye shadow, Bronzer)      (Mascara)    0.879433  2.463397
14                 (Mascara, Bronzer)   (Eye shadow)    0.905109  2.375615
15              (Mascara, Foundation)   (Eye shadow)    0.864583  2.269248
16               (Lip Gloss, Mascara)   (Eye shadow)    0.872928  2.291150
17             (Eye shadow, Lipstick)      (Mascara)    0.852713  2.388552
18                (Mascara, Lipstick)   (Eye shadow)    0.909091  2.386065
19             (Eye shadow, Eyeliner)      (Mascara)    0.829670  2.324007
20                (Mascara, Eyeliner)   (Eye shadow)    0.862857  2.264717
21     (Concealer, Eye shadow, Blush)      (Mascara)    0.959677  2.688172
22        (Concealer, Mascara, Blush)   (Eye shadow)    0.908397  2.384244
23  (Concealer, Eye shadow, Eyeliner)      (Mascara)    0.876923  2.456367
24     (Concealer, Mascara, Eyeliner)   (Eye shadow)    0.850746  2.232930
25   (Lip Gloss, Mascara, Foundation)   (Eye shadow)    0.853846  2.241066

## a. Interpret the first three rules in the output in words.
Interpreting the results of the association rules in the first three rules, we see that when brushes are purchased, there is a 100% confidence that nail polish will also be purchased with a lift ratio of 3.57. We see with rules 2 and 3 that there is some apparent redundancy as in rule 2 eye shadow is the antecedent and mascara is the consequent and in rule 3 we see the opposite, mascara is the antecedent and eye shadow is the consequent. Now, there is clearly a relationship between the two with confidence being 84.25% for rule 2 and the confidence being 89.92 for rule three with the same lift ratios in both cases. 

## b. Reviewing the first couple of dozen rules, comment on their redundancy and how you would assess their utility.
We've already commented on rules 2 and 3. Looking through our list of rules, we see that nearly every rule contains the relationship between eye shadow and mascara, combined with some other combination of products. This does not necessarily mean our rules are useless; however, we would need to assess our different rules and do some consolidating of rules before we could move forward with new data. Another notable mention is rules 9,10, and 11. These rules also contain the elements concealer, eyeliner, bronzer, and concealment which might also have the ability to be consolidated. 

## Course Ratings:
## Scenario: 
The Institute for Statistics Education at Statistics.com asks students to rate
a variety of aspects of a course as soon as the student completes it. The Institute is
contemplating instituting a recommendation system that would provide students with
recommendations for additional courses as soon as they submit their rating for a
completed course

## Question 3:
### Just reviewing the data visually, without doing much calculation, which user(s) would you consider most similar to E.N.?


In [96]:

course_data = pd.read_csv("courserating.csv")
course_data.set_index("Unnamed: 0",inplace=True)
course_data

SQL  Spatial  PA1  DM in R  Python  Forecast  R Prog  Hadoop  \
Unnamed: 0                                                                 
LN          4.0      NaN  NaN      NaN     3.0       2.0     4.0     NaN   
MH          3.0      4.0  NaN      NaN     4.0       NaN     NaN     NaN   
JH          2.0      2.0  NaN      NaN     NaN       NaN     NaN     NaN   
EN          4.0      NaN  NaN      4.0     NaN       NaN     4.0     NaN   
DU          4.0      4.0  NaN      NaN     NaN       NaN     NaN     NaN   
FL          NaN      4.0  NaN      NaN     NaN       NaN     NaN     NaN   
GL          NaN      4.0  NaN      NaN     NaN       NaN     NaN     NaN   
AH          NaN      3.0  NaN      NaN     NaN       NaN     NaN     NaN   
SA          NaN      NaN  4.0      NaN     NaN       NaN     NaN     NaN   
RW          NaN      NaN  2.0      NaN     NaN       NaN     NaN     4.0   
BA          NaN      NaN  4.0      NaN     NaN       NaN     NaN     NaN   
MG          NaN      NaN  4.0      NaN     NaN       4.0     NaN     NaN   
AF          NaN      NaN  4.0      NaN     NaN       NaN     NaN     NaN   
KG          NaN      NaN  3.0      NaN     NaN       NaN     NaN     NaN   
DS          4.0      NaN  NaN      2.0     NaN       NaN     4.0     NaN   

            Regression  
Unnamed: 0              
LN                 2.0  
MH                 NaN  
JH                 NaN  
EN                 3.0  
DU                 NaN  
FL                 NaN  
GL                 NaN  
AH                 NaN  
SA                 NaN  
RW                 NaN  
BA                 NaN  
MG                 NaN  
AF                 NaN  
KG                 NaN  
DS                 NaN

Based on the above data, from simply looking at the data, we see two entries that are similar to entry EN, LN and DS. LN and DS both have three class ratings in common with EN, albeit with different values for the ratings. We see DS and EN shared ratings with SQL (4.0), R Prog (4.0), and DM in R (here there is a difference of 2 in the rating values). Between LN and EN we see three classes with shared ratings as well: SQL (4.0), R Prog (4.0), and Regression (here there is a difference of 1). Since the total difference between LN and EN is less than the total distance between DS and EN, we can say that LN and EN are more similar. 

In [97]:
ratings = []
for customer,row in course_data.iterrows():
    for course, value in row.iteritems():
        if value==0: continue
        ratings.append([customer,course,value])
ratings = pd.DataFrame(ratings,columns=["customer","course","value"])
ratings.head()

customer   course  value
0       LN      SQL    4.0
1       LN  Spatial    NaN
2       LN      PA1    NaN
3       LN  DM in R    NaN
4       LN   Python    3.0

In [98]:
from surprise import KNNBasic

sim_options = {"name":"cosine","user_based":False}
algo = KNNBasic(sim_options=sim_options)
algo.fit(data.build_full_trainset())

ImportError: cannot import name 'KNNBasic' from 'surprise' (unknown location)

In [ ]:
predictions = []
for user in course_data.index:
    predictions.append([algo.predict(user,course).est for course in course_data])
predictions = pd.DataFrame(predictions,columns=course_data.columns)

NameError: name 'algo' is not defined

## Question 4:
### Use Python to compute the cosine similarity between users.

In [ ]:
from scipy.spatial.distance import cosine
def cosine_similarity_NA(data):
     m = data.shape[0]
     # Initialize the similarity matrix to np.nan
     result = np.full((m, m), np.nan)
     # Iterate over all pairs of columns
     for i in range(m):
         # maski is true if a value exists in column i
         maski = ~np.isnan(data.iloc[i])
         for j in range(i, m):
             # maskij is true if a value exists in both column i and j
             maskij = maski & ~np.isnan(data.iloc[j])
             if np.any(maskij):
                 # if there are values in both columns, calculate the cosine similarity
                 # for these
                 result[i, j] = 1 - cosine(data.iloc[i][maskij], data.iloc[j][maskij])
                 result[j, i] = result[i, j]
     return pd.DataFrame(result, columns=data.index, index=data.index)




cosine_similarity_NA(course_data)

Unnamed: 0      LN        MH        JH       EN        DU   FL   GL   AH   SA  \
Unnamed: 0                                                                      
LN          1.0000  0.960000  1.000000  0.98910  1.000000  NaN  NaN  NaN  NaN   
MH          0.9600  1.000000  0.989949  1.00000  0.989949  1.0  1.0  1.0  NaN   
JH          1.0000  0.989949  1.000000  1.00000  1.000000  1.0  1.0  1.0  NaN   
EN          0.9891  1.000000  1.000000  1.00000  1.000000  NaN  NaN  NaN  NaN   
DU          1.0000  0.989949  1.000000  1.00000  1.000000  1.0  1.0  1.0  NaN   
FL             NaN  1.000000  1.000000      NaN  1.000000  1.0  1.0  1.0  NaN   
GL             NaN  1.000000  1.000000      NaN  1.000000  1.0  1.0  1.0  NaN   
AH             NaN  1.000000  1.000000      NaN  1.000000  1.0  1.0  1.0  NaN   
SA             NaN       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
RW             NaN       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
BA             NaN       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
MG          1.0000       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
AF             NaN       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
KG             NaN       NaN       NaN      NaN       NaN  NaN  NaN  NaN  1.0   
DS          1.0000  1.000000  1.000000  0.96225  1.000000  NaN  NaN  NaN  NaN   

Unnamed: 0   RW   BA   MG   AF   KG       DS  
Unnamed: 0                                    
LN          NaN  NaN  1.0  NaN  NaN  1.00000  
MH          NaN  NaN  NaN  NaN  NaN  1.00000  
JH          NaN  NaN  NaN  NaN  NaN  1.00000  
EN          NaN  NaN  NaN  NaN  NaN  0.96225  
DU          NaN  NaN  NaN  NaN  NaN  1.00000  
FL          NaN  NaN  NaN  NaN  NaN      NaN  
GL          NaN  NaN  NaN  NaN  NaN      NaN  
AH          NaN  NaN  NaN  NaN  NaN      NaN  
SA          1.0  1.0  1.0  1.0  1.0      NaN  
RW          1.0  1.0  1.0  1.0  1.0      NaN  
BA          1.0  1.0  1.0  1.0  1.0      NaN  
MG          1.0  1.0  1.0  1.0  1.0      NaN  
AF          1.0  1.0  1.0  1.0  1.0      NaN  
KG          1.0  1.0  1.0  1.0  1.0      NaN  
DS          NaN  NaN  NaN  NaN  NaN  1.00000

Above is a matrix of the different cosine values from the ratings dataframe. If we wanted to observe the cosine for LN and EN, we could look at the cell that matches the two of them and return the value. The value is 0.9891. Looking across the row for EN, we observe that DS has a cosine relationship with EN of 0.9623. 

## Question 5: 
### Based on the cosine similarities of the nearest students to E.N., which course should be recommended to E.N.?

Looking at the cosine table and back at the course dataframe, we do see that only DS and Ln are closest to EN so we will see which classes they have taken that EN has not. We see that DS only took a total of three classes, so this does not help us recommend a new class to EN. However, looking at LN, we see LN has taken five classes, two of which EN has not taken- Python and Forecast. Based on this information and the fact that LN gave Python a higher rating, we should recommend Python first and then Forecast. 

## Question 6:
### With large datasets, it is computationally difficult to compute user-based recommendations in real time, and an item-based approach is used instead. Returning to the rating data (not the binary matrix), let’s now take that approach.
### 1. If the goal is still to find a recommendation for E.N., for which course pairs is it possible and useful to calculate correlations?

It seems there might be some sort of correlation between SQL and Spatial and potentially Python and SQL as there are some similar ratings between these courses. 

### 2. Just looking at the data, and without yet calculating course pair correlations, which course would you recommend to E.N., relying on itembased filtering? Calculate two course pair correlations involving your guess and report the results.

Without calculating course pair correlations, I would recommend Spatial and Python to user E.N. 

Below are the course pair correlations:
- Spatial and SQL: 0.8182 
- Python and SQL: -0.9999 or -1, a nearly perfect negative correlation

## Question 7: 
### Apply item-based collaborative filtering to the dataset (using the Surprise package in Python) and based on the results, recommend a course to E.N.

In [ ]:
user_id = "EN"
if user_id in ratings['customer'].unique():
    user_ratings = ratings[ratings['customer'] == user_id]
    predictions = [(course, algo.predict(user_id, course).est) for course in user_ratings['course']]

    print(f'Predictions for user {user_id}:')
    for course, prediction in predictions:
        print(f'Course: {course}, Prediction: {prediction}')

In the output above, we see the courses EN has already taken along with the other likelihoods of the other courses. Based on the prediction from the code, Python has the highest prediction likelihood of the courses EN has not yet taken at 3.5027. 